test/integration/schema/index.js

```diff
@@ -576,5 +576,111 @@ module.exports = function(knex) {
       });
     });
 
+    it('supports named primary keys', function() {
+      var constraintName = 'pk-test';
+      var tableName = 'namedpk';
+      return knex.transaction(function(tr) {
+        return tr.schema.dropTableIfExists(tableName)
+          .then(function() {
+            return tr.schema.createTable(tableName, function(table) {
+              table.string('test').primary(constraintName);
+            })
+          })
+          .then(function() {
+            return tr.schema.dropTableIfExists(tableName);
+          })
+          .then(function() {
+            return tr.schema.createTable(tableName, function(table) {
+              table.string('test');
+              table.string('test2');
+            })
+          })
+          .then(function() {
+            return tr.schema.table(tableName, function(table) {
+              table.primary('test', constraintName)
+            })
+          })
+          .then(function() {
+            return tr.schema.dropTableIfExists(tableName);
+          })
+          .then(function() {
+            return tr.schema.createTable(tableName, function(table) {
+              table.string('test');
+              table.string('test2');
+            })
+          })
+          .then(function() {
+            return tr.schema.table(tableName, function(table) {
+              table.primary(['test', 'test2'], constraintName)
+            })
+          });
+      });
+    });
+
+    it('supports named unique keys', function() {
+      var singleUniqueName = 'uk-single';
+      var multiUniqueName = 'uk-multi';
+      var tableName = 'nameduk';
+      return knex.transaction(function(tr) {
+        return tr.schema.dropTableIfExists(tableName)
+          .then(function() {
+            return tr.schema.createTable(tableName, function(table) {
+              table.string('test').unique(singleUniqueName);
+            })
+          })
+          .then(function() {
+            return tr.schema.dropTableIfExists(tableName);
+          })
+          .then(function() {
+            return tr.schema.createTable(tableName, function(table) {
+              table.string('test');
+              table.string('test2');
+            })
+          })
+          .then(function() {
+            return tr.schema.table(tableName, function(table) {
+              table.unique('test', singleUniqueName);
+              table.unique(['test', 'test2'], multiUniqueName);
+            })
+          });
+      });
```

Comment:
maybe creating other constraint with the same name could also work for testing...